MaxText/max_utils.py

```diff
@@ -230,20 +230,51 @@ def setup_initial_state(model, tx, config, rng, mesh, checkpoint_manager):
 
 # Learning Rate Schedule
 # -----------------------------------------------------------------------------
-# learning rate scheduling
-def rsqrt_schedule(init_value: float, shift: int = 0):
-  def schedule(count):
-    return init_value * (1 + count + shift)**-.5 * shift**.5
-  return schedule
-
-def create_learning_rate_schedule(learning_rate: float, warmup_steps: int):
-  """Creates a rsqrt schedule with linear warmup."""
-  return optax.join_schedules([
-      optax.linear_schedule(
-          init_value=0,
-          end_value=learning_rate,
-          transition_steps=warmup_steps
-          ),
-      rsqrt_schedule(init_value=learning_rate, shift=warmup_steps),
-      ], boundaries=[warmup_steps],
-      )
+
+def create_learning_rate_schedule(config):
+  """Creates a four part schedule:
+  1) Linear warmup from 0 to [learning_rate]
+  2) Cosine from [learning_rate] to [learning_rate * cosine_learning_rate_final_fraction]
+  3) Linear cooldown from [learning_rate * cosine_learning_rate_final_fraction] to 0
+  4) Constant learning rate of 0 
+  The constant zero learning at the rate aids in both accurately measuring the loss of the final model
+  over a large set of examples, and for debugging:the loss should remain roughly constant during this regime.
+  """
+  def make_cos_schedule(init_lr, final_lr, len_steps):
+    def schedule(step):
+      pct = (step) / len_steps
+      a = 0.5 * (jnp.cos(jnp.pi*pct) + 1)
+      lr = init_lr * a + final_lr * (1 - a)
+      return lr
+    return schedule
+
+  steps = config.steps
```

Comment:
We need 2 params in config:
config.schedule_steps # this one is controlling when the LR schedule ends (goes to zero after cooldown)
config.run_steps # this one is controlling when the training should finish - if we want to finish early or if we want to train longer to get 0 LR for some more steps to get good final measurement.
If either one is -1, it should inherit value from the other one (just like in int8-sweeps) branch.